page-factory-core/src/main/java/ru/sbtqa/tag/pagefactory/data/DataParser.java

```diff
@@ -0,0 +1,183 @@
+
+    private String featureDataTag;
+    private String currentScenarioTag;
+
+    public void replaceDataPlaceholders(List<CucumberFeature> cucumberFeatures) throws DataException, IllegalAccessException {
+
+        for (CucumberFeature cucumberFeature : cucumberFeatures) {
+            GherkinDocument gherkinDocument = cucumberFeature.getGherkinFeature();
+            Feature feature = gherkinDocument.getFeature();
+
+            setFeatureDataTag(parseTags(feature.getTags()));
+            List<ScenarioDefinition> featureChildren = feature.getChildren();
+
+            for (ScenarioDefinition scenarioDefinition : featureChildren) {
+                List<Tag> currentScenarioTags = getScenarioTags(scenarioDefinition);
+                setCurrentScenarioTag(parseTags(currentScenarioTags));
+                List<Step> steps = scenarioDefinition.getSteps();
+
+                if (scenarioDefinition instanceof ScenarioOutline) {
+                    List<Examples> examples = ((ScenarioOutline) scenarioDefinition).getExamples();
+                    FieldUtils.writeField(scenarioDefinition, "examples", replaceExamplesPlaceholders(examples), true);
+                }
+
+                for (Step step : steps) {
+                    FieldUtils.writeField(step, "argument", replaceArgumentPlaceholders(step.getArgument()), true);
+                    FieldUtils.writeField(step, "text", replaceDataPlaceholders(step.getText()), true);
+                }
+            }
+        }
+    }
+
+    private List<Tag> getScenarioTags(ScenarioDefinition scenarioDefinition) {
+        try {
+            return (List<Tag>) FieldUtils.readField(scenarioDefinition, "tags", true);
+        } catch (IllegalArgumentException | IllegalAccessException e) {
+            return new ArrayList<>();
+        }
+    }
+
+    private String parseTags(List<Tag> tags) {
+        Optional<Tag> dataTag = tags.stream().filter(predicate -> predicate.getName().startsWith("@data")).findFirst();
+        return dataTag.isPresent() ? dataTag.get().getName().split("=")[1].trim() : null;
+    }
+
+    private Node replaceArgumentPlaceholders(Node argument) throws DataException {
+        if (argument instanceof DataTable) {
+            DataTable dataTable = (DataTable) argument;
+            argument = replaceDataTablePlaceholders(dataTable);
+        } else if (argument instanceof DocString) {
+            DocString docString = (DocString) argument;
+            argument = new DocString(docString.getLocation(), docString.getContentType(), replaceDataPlaceholders(docString.getContent()));
+
+        }
+        return argument;
+    }
+
+    private String replaceDataPlaceholders(String raw) throws DataException {
+        Pattern stepDataPattern = Pattern.compile(STEP_PARSE_REGEX);
+        Matcher stepDataMatcher = stepDataPattern.matcher(raw);
+        StringBuffer replacedStep = new StringBuffer(raw);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace the synchronized class "StringBuffer" by an unsynchronized one such as "StringBuilder". ([squid:S1149](https://sonarcloud.io/organizations/sbtqasonarkey/rules#rule_key=squid%3AS1149))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=ru.sbtqa.tag%3Apage-factory-2&pullRequest=64&issues=AWSn0wibsk71k7lw7bXR&open=AWSn0wibsk71k7lw7bXR)

[ISSUE_KEY]: AWSn0wibsk71k7lw7bXR